lightning/src/lib.rs

```diff
@@ -28,31 +28,132 @@
 
+#[cfg(not(any(feature = "std", feature = "no_std")))]
+compile_error!("at least one of the `std` or `no_std` features must be enabled");
+
 #[macro_use]
 extern crate alloc;
 extern crate bitcoin;
+#[cfg(any(test, feature = "std"))]
 extern crate core;
+
 #[cfg(any(test, feature = "_test_utils"))] extern crate hex;
 #[cfg(any(test, feature = "fuzztarget", feature = "_test_utils"))] extern crate regex;
 
+#[cfg(feature = "no_std")] extern crate core2;
+
+
 #[macro_use]
 pub mod util;
 pub mod chain;
 pub mod ln;
 pub mod routing;
 
+#[cfg(feature = "std")]
+use std::io;
+#[cfg(not(feature = "std"))]
+use core2::io;
+
+#[cfg(not(feature = "std"))]
+mod io_extras {
+	use core2::io::{self, Read, Write};
+
+	/// A writer which will move data into the void.
+	pub struct Sink {
+		_priv: (),
+	}
+
+	/// Creates an instance of a writer which will successfully consume all data.
+	pub const fn sink() -> Sink {
+		Sink { _priv: () }
+	}
+
+	impl core2::io::Write for Sink {
+		#[inline]
+		fn write(&mut self, buf: &[u8]) -> core2::io::Result<usize> {
+			Ok(buf.len())
+		}
+
+		#[inline]
+		fn flush(&mut self) -> core2::io::Result<()> {
+			Ok(())
+		}
+	}
+
+	pub fn copy<R: ?Sized, W: ?Sized>(reader: &mut R, writer: &mut W) -> Result<u64, io::Error>
+		where
+		R: Read,
+		W: Write,
+	{
+		let mut count = 0;
+		let mut buf = [0u8; 64];
+
+		loop {
+			match reader.read(&mut buf) {
+				Ok(0) => break,
+				Ok(n) => { writer.write_all(&buf[0..n])?; count += n as u64; },
+				Err(ref e) if e.kind() == io::ErrorKind::Interrupted => {},
+				Err(e) => return Err(e.into()),
+			};
+		}
+		Ok(count)
+	}
+
+	pub fn read_to_end<D: io::Read>(mut d: D) -> Result<alloc::vec::Vec<u8>, io::Error> {
+		let mut result = vec![];
+		let mut buf = [0u8; 64];
+		loop {
+			match d.read(&mut buf) {
+				Ok(0) => break,
+				Ok(n) => result.extend_from_slice(&buf[0..n]),
+				Err(ref e) if e.kind() == io::ErrorKind::Interrupted => {},
+				Err(e) => return Err(e.into()),
+			};
+		}
+		Ok(result)
+	}
+}
+
+#[cfg(feature = "std")]
+mod io_extras {
+	pub fn read_to_end<D: ::std::io::Read>(mut d: D) -> Result<Vec<u8>, ::std::io::Error> {
+		let mut buf = Vec::new();
+		d.read_to_end(&mut buf)?;
+		Ok(buf)
+	}
+}
+
 mod prelude {
 	#[cfg(feature = "hashbrown")]
 	extern crate hashbrown;
 
-	pub use alloc::{vec, vec::Vec, string::String, collections::VecDeque};
+	pub use alloc::{vec, vec::Vec, string::String, collections::VecDeque, boxed::Box};
 	#[cfg(not(feature = "hashbrown"))]
 	pub use std::collections::{HashMap, HashSet, hash_map};
 	#[cfg(feature = "hashbrown")]
 	pub use self::hashbrown::{HashMap, HashSet, hash_map};
+
+	#[cfg(feature = "std")]
+	pub use std::io::sink;
+
+	#[cfg(not(feature = "std"))]
+	pub use io_extras::sink;
+
+	#[cfg(feature = "std")]
+	pub use std::io::copy;
+
+	#[cfg(not(feature = "std"))]
+	pub use io_extras::copy;
+
+	pub use alloc::borrow::ToOwned;
+	pub use alloc::string::ToString;
+
+	pub use io_extras::read_to_end;
```

Comment:
Just personal preference, but I'm really not a fan of transparently including functions in prelude - it makes it somewhat unclear where they're coming from in code using it. Can we instead export `copy` from `io_extras` and then explicitly use `copy` and `read_to_end` from `io_extras` in files that use them?

Reply:
Agreed.  Moving those and also `sink`.